client-side-virtual-keyboard.md

```diff
@@ -0,0 +1,229 @@
+        *
+        * @param start the position in the text at which the composing region begins.
+        * @param end the position in the text at which the composing region ends.
+        */
+       @Override
+       public void onVirtualKeyboardComposingTextRegionChanged(int start, int end) {
+       }
+
+       /**
+        * Called as the state of the virtual keyboard is changed.
+        *
+        * @param visible the visibility of the virtual keyboard, true or false.
+        * @param heightRatio the ratio of virtual keyboard's height to screen when virtual keyboard is visible.
+        *        This can be used to notify the IME running in the Android container
+        *        to adjust the display height to honor the virtual keyboard display
+        *        ratio on the client side.
+        *
+        */
+       @Override
+       public void onVirtualKeyboardStateChanged(boolean visible, double displayRatio) {
+       }
+
+   ```
+
+6. In HTML, Defines the JavaScript functions that can be invoked from Android Java layer(AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK. This builds up a communication tunnel that works the other way around.
+   ```
+   <script>
+     function sendIMECommittedText(text) {
+       window.stream.sendIMECommittedText(text);
+     }
+     function sendIMEComposingText(text) {
+       window.stream.sendIMEComposingText(text);
+     }
+     function sendIMETextDeletion(counts) {
+       window.stream.sendIMETextDeletion(counts);
+     }
+     function sendIMEAction(name, params) {
+       window.stream.sendIMEAction(name, params);
+     }
+     function sendIMEComposingRegion(start, end) {
+       window.stream.sendIMEComposingRegion(start, end);
+     }
+   </script>
+
+   ```
+
+7. Once the virtual keyboard pops up on the client side, the `onVirtualKeyboardStateChanged` callback function will be triggered, to keep display proportions correct for IME displaying on both client and server ends, the action with the name `show` that carries display ratio needs to be sent out to the server side.
+
+   ```
+       @Override
+       public void onVirtualKeyboardStateChanged(boolean visible, double displayRatio) {
+           if (visible == true) {
+               String action = "show";
+               String params = "height-ratio=" + displayRatio;
+               mWebView.loadUrl(String.format("javascript:sendIMEAction(\"%s\", \"%s\")", action, params));
+           }
+       }
+
+   ```
+
+8. While typing on the virtual keyboard, a text input event will be triggered when one of the following scenarios occurs:
+   - A bunch of characters is committed to a text editor
+   - A text is currently in composing
+   - A text is being deleted
+
+    In the above cases, the changed text must be sent to the server side through the JS SDK. This can be done by calling JS functions that are defined in HTML through AnboxWebView. E.g. for the case of text commitment.
```

Comment:
```suggestion
    In the above cases, the changed text must be sent to the server side through the JavaScript SDK. This can be done by calling JavaScript functions that are defined in HTML through AnboxWebView. For example, for committing text:
```

Reply:
+1, Will fix